### PR TITLE
feat: RP workspace mapping SDK [DET-9522]

### PR DESCRIPTION
**Pull request by jgongd**
## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

[DET-9522]

[DET-9522]: https://hpe-aiatscale.atlassian.net/browse/DET-9522?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: harness/determined/common/experimental/determined.py
Comment (eecsliu) [inline harness/determined/common/experimental/determined.py:408]: does a function like this already exist? I'd be surprised if there wasn't something for this

File: harness/determined/common/experimental/determined.py
Comment (eecsliu) [inline harness/determined/common/experimental/determined.py:434]: I believe we only want to bind one resource pool at a time, to keep it consistent with the API and webui

File: harness/determined/experimental/client.py
Comment (eecsliu) [inline harness/determined/experimental/client.py:454]: same comment as above, I think we want to only bind to one resource pool at a time

**Review by eecsliu — COMMENTED** [review]
It looks mostly good to me, save one function argument type. I would also wait for a review by someone from MLSys

File: harness/determined/common/experimental/determined.py
Comment (jgongd) [reply harness/determined/common/experimental/determined.py:434]: Cool, that makes more sense. I will update the PRD.

File: harness/determined/experimental/client.py
Comment (jgongd) [inline harness/determined/experimental/client.py:471]: Same as above. We want to only unbind one resource pool at a time.

File: harness/determined/common/experimental/determined.py
Comment (wes-turner) [inline harness/determined/common/experimental/determined.py:447]: One of the principles we're reinforcing in the SDK is to keep the interface simple, and not do work that our users can easily do themselves. Here, that means
```python
def bind_rp_to_workspace(self, resource_pool_name: str, workspace_name: List[str]) -> None:
```
rather than
```python
def bind_rps_to_workspace(self, resource_pool_names: List[str], workspace_name: List[str]) -> None:
```

I'd ideally also like it even more if we could even better
```python
def bind_rp_to_workspace(self, resource_pool_name: str, workspace_name: str) -> None:
```
but I get it if product has latency targets for a use case this wouldn't perform well for.

File: harness/determined/common/experimental/determined.py
Comment (wes-turner) [inline harness/determined/common/experimental/determined.py:408]: Just a short note that if we bound resource pools to workspaces by names, we wouldn't need this function.

File: harness/determined/common/experimental/determined.py
Comment (wes-turner) [inline harness/determined/common/experimental/determined.py:481]: To fit with the rest of the SDK, I'd prefer an interface with the following translations:
* `list_rps_bound_to_workspace` -> `Workspace.resource_pools()`
* `bind_rps_to_workspaces` -> `ResourcePool.bind(workspace_ids)`
* `list_workspaces_bound_to_rp` -> `ResourcePool.workspaces()`
* `overwrite_rp_workspace_bindings` -> `ResourcePool.set_bindings(workspace_ids)`

Currently neither Workspace nor ResourcePool exist (and it's not even clear what a "ResourcePool" should represent). If you approve of this interface, I'd be happy to create them in another PR.

File: harness/determined/common/experimental/determined.py
Comment (wes-turner) [inline harness/determined/common/experimental/determined.py:432]: Per an OOB thread, there's not much in this PR that wants a unit test. Most of the functionality is in assembling parameters to be passed through to the REST API, and though we _could_ write unit tests that the requests are correct, I don't myself think that's necessary.

But nevertheless, a set of unit tests for this method could include:

* `test_bind_rp_to_workspaces_errors_on_invalid_rps`
* `test_bind_rp_to_workspaces_errors_on_invalid_workspaces`

`list_rps_bound_to_workspace` could test:
* `test_list_rps_bound_to_workspace_reads_all_pages`
* `test_list_rps_bound_to_workspace_raises_NotFound_on_invalid_workspace`
* `test_list_rps_bound_to_workspace_returns_empty_list_when_no_rps_are_bound`

Comment (wes-turner) [reply harness/determined/common/experimental/determined.py:432]: In case either is helpful, there's a [page in confluence](https://hpe-aiatscale.atlassian.net/wiki/spaces/ENGINEERIN/pages/1351909582/Python+Testing) with some best practices for unit tests, and [examples already in the SDK](https://github.com/determined-ai/determined/tree/main/harness/tests/determined/common/experimental) to crib from.

File: master/internal/db/postgres_rp_workspace_bindings_intg_test.go
Comment (jgongd) [inline master/internal/db/postgres_rp_workspace_bindings_intg_test.go:1]: I don't expect this. I probably made a mistake in rebasing.

File: harness/determined/common/experimental/workspace.py
Comment (wes-turner) [inline harness/determined/common/experimental/workspace.py:60]: nit: because I expect this returns a list no matter if there are bound resource pools or not, I think the best return type is just `List[str]`.

File: harness/determined/common/experimental/resource_pool.py
Comment (wes-turner) [inline harness/determined/common/experimental/resource_pool.py:39]: We've been inconsistent within the codebase and I don't know how you could have known otherwise, but please follow the Google style guide (itself a more detailed specification than PEP-8) [for docstrings](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods).

It's a lot of text to read, but it works pretty well to just crib off the examples.

File: harness/determined/common/experimental/resource_pool.py
Comment (wes-turner) [inline harness/determined/common/experimental/resource_pool.py:36]: The paths within the CLI have recently undergone an evolution, and I think it could make sense to mirror them here with `add_binding`, `remove_binding`, and `replace_binding`. But I also like this interface, too.

Just wanted to bring in the most up-to-date information possible so you could make whatever decision you most prefer. I think both sets of names are great.

File: harness/determined/common/experimental/resource_pool.py
Comment (wes-turner) [inline harness/determined/common/experimental/resource_pool.py:65]: I think there's a small bug here with `workspaceIds` <> `workspaceNames`.

File: harness/determined/common/experimental/resource_pool.py
Comment (wes-turner) [inline harness/determined/common/experimental/resource_pool.py:51]: I haven't spent as much time with `bindings` as you'd expect, but it seems to me that some functions take just straight parameters, and some take a constructed request. The former seems more user-friendly (and I think is more common? Am unsure).

What's the motivation behind first creating a request and then sending that request, here?

side note: whatever the answer, it's a little weird that the `resourcePoolName` is needed in both.

File: harness/determined/common/experimental/resource_pool.py
Comment (wes-turner) [inline harness/determined/common/experimental/resource_pool.py:57]: A description in the docstring about what happens in edge cases (like if the workspaces don't resolve to names, if they're already unbound, etc.) would be helpful to a caller.

File: harness/determined/common/experimental/resource_pool.py
Comment (wes-turner) [inline harness/determined/common/experimental/resource_pool.py:89]: This conditional is surprising to me. When does it happen?

File: harness/determined/common/experimental/resource_pool.py
Comment (wes-turner) [reply harness/determined/common/experimental/resource_pool.py:57]: `bind` and `set` have fewer edge cases, but I think still warrant a "what happens when the name isn't found" description.

File: harness/tests/determined/common/experimental/test_resource_pool.py
Comment (wes-turner) [inline harness/tests/determined/common/experimental/test_resource_pool.py:27]: I'm going to have to think about this one a little longer (got another meeting right now). I'm not sure what behavioral contracts we want to guarantee with this test.

File: harness/tests/determined/common/experimental/test_workspace.py
Comment (wes-turner) [inline harness/tests/determined/common/experimental/test_workspace.py:92]: I think the behavior under test here looks like `test_resource_pools_gets_single_resource_pool` or the like. You've set up the fixtures and managed `responses` well, but I'm not sure there's enough logic in the method itself for this specific behavior that it wants a unit test.

I'll need to come back to this after my next meeting, here. I think the key will be to think about the code path under exercise with this test for the code that's specifically in `Workspace.respource_pools`.

File: harness/tests/determined/common/experimental/test_resource_pool.py
Comment (wes-turner) [reply harness/tests/determined/common/experimental/test_resource_pool.py:27]: First (and the easiest thing to write about and address), if the test is specifically testing that an `APIException` is raised, it should be done via `with pytest.raises...`.

Comment (wes-turner) [reply harness/tests/determined/common/experimental/test_resource_pool.py:27]: Different behaviors we might want to guarantee (from behavior we have control over at this layer):
1. The `ResourcePool` constructor will throw an exception rather than create an object if an invalid name is given
2. `bind`, `unbind`, etc. will throw an exception if the bindings call constructed from them raises an exception.

I'm not sure we can do much else.

For (1), we don't currently do this now, but if you'd like to add this functionality I'd support it. And if you do add it, I think it deserves a unit test.

For (2), I don't think "don't catch exceptions passed by the bindings" is behavior worthy of a unit test. But I have a vague sense I'm not seeing the entire picture here, so if you see this different, let's talk about it in text/huddle.

File: harness/determined/common/experimental/resource_pool.py
Comment (eecsliu) [reply harness/determined/common/experimental/resource_pool.py:51]: yeah this is the first time I'm working with bindings too, so I'm not sure why some are requests and others are parameters. It seems like requests are used when there are more than one parameter to pass? The `resourcePoolName` parameter here could be translated from the proto, which has `resourcePoolName` as part of the URL and as part of the request

File: harness/determined/common/experimental/resource_pool.py
Comment (eecsliu) [reply harness/determined/common/experimental/resource_pool.py:89]: I think this is because resps can be more than one response due to pagination, so we have to check each response to make sure workspaceIds is populated?

File: harness/tests/determined/common/experimental/test_resource_pool.py
Comment (eecsliu) [reply harness/tests/determined/common/experimental/test_resource_pool.py:27]: hmm right now, there isn't an easy, RBAC-ful way to check if the resource pool is valid on startup, so I'm inclined to go with option 2. If that's the case, I'm ok removing this test

Comment (wes-turner) [reply harness/tests/determined/common/experimental/test_resource_pool.py:27]: I'm in -- let's kill it.

File: harness/tests/determined/common/experimental/test_workspace.py
Comment (wes-turner) [reply harness/tests/determined/common/experimental/test_workspace.py:92]: The `resource_pools` function is so simple -- as syntactic sugar on `bindings`, it mostly just passes arguments through. The only part of it that has any trickiness for me is the pagination part. That part is hard to think about.

So I think to the extent you want to include any unit test for this method, testing that it actually does compile results across multiple pages is the only code I think is worth a unit test.

If this was my PR, I'd like to think that I'd include a test for it. But when the rubber hits the road I might end up being too lazy for it. Let's call it "recommended" but not "required."

If you want to, I've already done something very similar in [test_model.py](https://github.com/determined-ai/determined/blob/497df171c14ea3089925ba2880e273a9a5274f04/harness/tests/determined/common/experimental/test_model.py#L30) (added after discovering a bug in `get_versions`'s pagination).

**Review by wes-turner — APPROVED** [review]

File: harness/determined/common/experimental/resource_pool.py
Comment (jgongd) [reply harness/determined/common/experimental/resource_pool.py:113]: I put the check there for two reasons.
1. The Optional type you mentioned above.
2. Being conservative, I'm not sure what it returns if there is no workspace bound to the RP. An empty list or None? I should have probably played with it and read the docs.

Comment (jgongd) [reply harness/determined/common/experimental/resource_pool.py:113]: @wes-turner Good catch. The optional is due a misconfig in the proto file.
```
message ListWorkspacesBoundToRPResponse {
  option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
    json_schema: { required: [ "workspace_id", "limit" ] }
  };
  // List of workspace IDs.
  repeated int32 workspace_ids = 1;
  // Pagination information of the full dataset.
  Pagination pagination = 2;
}
```
```
message GetExperimentsResponse {
  option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
    json_schema: { required: [ "experiments", "pagination" ] }
  };
  // The list of returned experiments.
  repeated determined.experiment.v1.Experiment experiments = 1;
  // Pagination information of the full dataset.
  Pagination pagination = 2;
}
```

File: harness/determined/common/experimental/resource_pool.py
Comment (jgongd) [reply harness/determined/common/experimental/resource_pool.py:61]: I'm also new to this. Just to share my findings in case people also want to know. It's due to how we define proto.

Payload definition:
```
message BindRPToWorkspaceRequest {
  option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
    json_schema: { required: [ "resource_pool_name" ] }
  };

  // The resource pool name.
  string resource_pool_name = 1;
  // The workspace IDs to be bound to the resource pool.
  repeated int32 workspace_ids = 2;
  // The workspace names to be bound to the resource pool.
  repeated string workspace_names = 3;
}
```
API endpoint definition:
```
  rpc BindRPToWorkspace(BindRPToWorkspaceRequest)
      returns (BindRPToWorkspaceResponse) {
    option (google.api.http) = {
      post: "/api/v1/resource-pools/{resource_pool_name}/workspace-bindings",
      body: "*"
    };
    option (grpc.gateway.protoc_gen_swagger.options.openapiv2_operation) = {
      tags: "Internal"
    };
  }
```

Comment (eecsliu) [reply harness/determined/common/experimental/resource_pool.py:61]: Yeah I'm not sure this is worth investigating, but perhaps something to look into for a tech debt ticket? Does seem interesting

**Pull request merged**
